File: src/main/java/arcana/common/research/BookEntry.java
package arcana.common.research;

import com.google.common.collect.ImmutableList;
import net.minecraft.item.Item;

public class BookEntry {
    public Item[] icons;
    public byte x, y;
    public BookSection[] sections; //includes requirements
    public ImmutableList<BookEntry> parents;
}

File: src/main/kotlin/arcana/common/research/BookEntry.kt
package arcana.common.research

import com.google.common.collect.ImmutableList
import net.minecraft.item.Item

class BookEntry {
    lateinit var icons: Array<Item>
    var x: Byte = 0
    var y: Byte = 0
    lateinit var sections: Array<BookSection>  //includes requirements
    lateinit var parents: ImmutableList<BookEntry>
}